atoma-p2p/src/service.rs

```diff
@@ -0,0 +1,1125 @@
+    /// * `message_data` - Raw bytes of the received gossipsub message (CBOR encoded)
+    /// * `message_id` - Unique identifier for the gossipsub message
+    /// * `propagation_source` - The peer that forwarded us this message
+    ///
+    /// # Returns
+    ///
+    /// Returns `Ok(())` if message processing succeeds, or an error variant if any step fails.
+    ///
+    /// # Errors
+    ///
+    /// This function can return several error types:
+    /// * `GossipsubMessageDataParseError` - If CBOR message deserialization fails
+    /// * `SignatureVerificationError` - If signature verification fails for an address response
+    /// * `GossipsubMessageRebroadcastError` - If rebroadcasting the message fails
+    /// * `SignHashedMessageError` - If signing the response message fails
+    ///
+    /// # Security Considerations
+    ///
+    /// - Messages are cryptographically signed to prevent tampering
+    /// - Timestamps are included to prevent replay attacks
+    /// - Invalid signatures cause messages to be dropped without rebroadcast
+    /// - Uses blake3 for secure message hashing
+    /// - Uses CBOR for efficient binary serialization
+    ///
+    /// # Message Format
+    ///
+    /// The messages are serialized using CBOR, but here's a JSON representation for readability:
+    ///
+    /// ```json
+    /// // Address Response
+    /// {
+    ///     "address": "node1.example.com:8080",
+    ///     "signature": [bytes],
+    ///     "timestamp": 1234567890,
+    ///     "node_small_id": 42
+    /// }
+    /// ```
+    ///
+    /// The actual wire format uses CBOR encoding for more efficient binary representation
+    /// and better handling of binary data like signatures.
+    #[instrument(level = "debug", skip_all)]
+    pub async fn handle_gossipsub_message(
+        &mut self,
+        message_data: &[u8],
+        message_id: &gossipsub::MessageId,
+        propagation_source: &PeerId,
+    ) -> Result<(), AtomaP2pNodeError> {
+        debug!(
+            target = "atoma-p2p",
+            event = "gossipsub_message",
+            message_id = %message_id,
+            propagation_source = %propagation_source,
+            "Received gossipsub message"
+        );
+        if propagation_source == self.swarm.local_peer_id() {
+            trace!(
+                target = "atoma-p2p",
+                event = "gossipsub_message_from_self",
+                "Gossipsub message from self"
+            );
+            // Do not re-publish the node's own message, just return `Ok(())
+            return Ok(());
+        }
+        let gossip_message: GossipMessage = serde_cbor::from_slice(message_data)?;
+        match gossip_message {
+            GossipMessage::AddressResponse(response) => {
+                debug!(
+                    target = "atoma-p2p",
+                    event = "gossipsub_message_data",
+                    "Received gossipsub message data"
+                );
+                let message_acceptance =
```

Comment:
@sdbondi I added here error handling for message validation, and providing the acceptance criteria resulting from it, which is then passed below in the `report_message_validation_result`. How does this look like ?